### PR TITLE
Feat. Video based tracking

### PR DESCRIPTION
**Pull request by Varun-sai-500**
This PR focuses on allowing videos for tracking purpose, I have added explicit flags such that it takes in a video path we specify along with the mask we provide at the beginning, then it takes frames out of it, and then carry on the rest of the pipeline. It runs, tested it.

 python tools/demo.py --video inpput.mp4 --mask 00002058.png  --fps 30 --ckpt_path "pretrain_models/DeAOTT_PRE_YTB_DAV.pth" --model deaott
 
 Edit: @z-x-yang The device logic will be clean once the MPS PR lands and both of the merges should definitely work because similar logic and already actually tested, so absolutely no doubt on that device flow, once the mps pr merges, because this depends on that

### CODE REVIEW

**Comment by Varun-sai-500** [comment]
@z-x-yang  Could you merge this? It's tested

**Review by z-x-yang — CHANGES_REQUESTED** [review]
Thanks for splitting this out.

I’m requesting changes for now because I still see a couple of blockers in the current implementation:

1. The new MPS path still accesses `torch.backends.mps` directly, so it has the same older-PyTorch compatibility problem as the dedicated MPS PR.
2. The device flow looks inconsistent in the new video path: parts of inference move tensors to MPS, but the checkpoint / model-loading path still looks CPU-oriented in this version. Please verify that model weights and input tensors are always on the same device end-to-end.
3. `place_and_validate_mask()` normalizes masks to binary `{0,1}`. That is fine for single-object masks, but it will collapse labels for multi-object masks. If multi-object inputs are supposed to work here, this needs to preserve indexed labels instead of binarizing them.

Once those are cleaned up, I’m happy to re-review.

**Comment by Varun-sai-500** [comment]
@z-x-yang  The device logic will be clean once the MPS PR lands and this PR changes along with that will definitely work because similar logic and already actually tested, so no doubt on that device flow.

Now this fixes the mask issue, multi-object masks are working fine!

**Review by z-x-yang — CHANGES_REQUESTED** [review]
Thanks for the update. The indexed-mask issue looks much better now: `place_and_validate_mask()` preserves the mask labels instead of binarizing everything, and the new demo-side MPS check is guarded.

I still cannot merge this PR as-is because the device-flow blocker is still present in this branch. In `tools/demo.py`, the model is initially moved to `device`, but then checkpoint loading calls `load_network(..., -1)` for non-CUDA. In this PR's `utils/checkpoint.py`, non-CUDA still resolves to CPU, so the loaded model can end up on CPU while the demo later moves `current_img` / `current_label` to the selected MPS device.

Please either rebase/include the checkpoint-side MPS handling from the dedicated MPS PR, or otherwise make sure model weights, engine state, and input tensors are on the same device end-to-end. After that, please smoke-test the video path again.

**Review by z-x-yang — APPROVED** [review]
Thanks for the update. I rechecked the MPS/device path; the guarded MPS detection, `to_device()` helper, indexed mask handling, and checkpoint/model device placement address my prior blockers. Approving.
